### PR TITLE
ARROW-4376: [Rust] Implement from_buf_reader for csv::Reader

### PR DESCRIPTION
**Pull request by paddyhoran**
Update to allow the csv reader to work with anything that implements the `Read` trait.  

Also adds a new constructor `from_buf_reader`.  Although 90% of use cases will just read a csv file on disk there are some situations where to need to pre-process or otherwise adjust the data before sending it to the csv reader.

### CODE REVIEW

**Comment by paddyhoran** [comment]
Forgot to mention that I did not update the csv builder, it still only works with `File`'s.  Some of the code that performs type inference uses methods specific to `File`'s.  This can be addressed in a follow up JIRA.

**Review by nevi-me — APPROVED** [review]
LGTM, might need another `fmt` after #3501 lands with https://github.com/apache/arrow/blob/db89566efcef2a2bee1b9411143123261554cf46/rust/rustfmt.toml

**Review by andygrove — APPROVED** [review]
Nice. Thanks @paddyhoran

**Pull request closed without merging**
